Docs/diataxis 2 (#898)
iotaledger/wasp: documentation/docs/guide/core_concepts/consensus.md

---

> description: IOTA Smart Contracts Consensus

Reviewer: The description could use a couple more words, this seems more like a title.

image: /img/logo/WASP_logo_dark.png
keywords:
- smart contracts
- consensus
- explanation
---
# Consensus

> To update the chain, its committee needs to reach a consensus, meaning that more than two thirds of its validators have to agree to change the state in the exact same way. This prevents a single malicious node from wrecking havoc over the chain, of course, but there are also more mundane reasons for individual nodes to act up.

Reviewer: ```suggestion
To update the chain, its committee needs to reach a consensus, meaning that more than two thirds of its validators have to agree to change the state in the exact same way. This prevents a single malicious node from wreaking havoc over the chain, but there are also more mundane reasons for individual nodes to act up.
```

> Smart contracts are deterministic, so all honest nodes will produce equal results — but only if the input was the same. Each validator node has its own access to The Tangle, so it may look a bit different to different nodes, as fresh transactions take time to propagate through the network. Validator nodes will receive smart contract requests with random delays in a random order, and, finally, all computers run on their own always slightly skewed clocks.

Reviewer: ```suggestion
Smart contracts are deterministic, so all honest nodes will produce identical results if the input is the same. Each validator node has its own connection to The Tangle. This means each validator node may have a slightly different perception of the Tangle, as new transactions take time to propagate through the network. Validator nodes will receive smart contract requests with random delays in a random order, and, finally, all computers run on their own always slightly skewed clocks.
```

> Smart contracts are deterministic, so all honest nodes will produce equal results — but only if the input was the same. Each validator node has its own access to The Tangle, so it may look a bit different to different nodes, as fresh transactions take time to propagate through the network. Validator nodes will receive smart contract requests with random delays in a random order, and, finally, all computers run on their own always slightly skewed clocks.

Reviewer 2: ```suggestion
Smart contracts are deterministic, so all honest nodes will produce equal results — but only if the input was the same. Each validator node has its own access to the Tangle, so it may look a bit different to different nodes, as fresh transactions take time to propagate through the network. Validator nodes will receive smart contract requests with random delays in a random order, and, finally, all computers run on their own always slightly skewed clocks.
```

## Batch Proposals

> As the first step, each node provides its own vision, a *batch proposal*. It contains a local timestamp, a list of unprocessed requests, and the node's partial signature of the current state's hash.

Reviewer: ```suggestion
At first, each node provides its vision, a *batch proposal*. It contains a local timestamp, a list of unprocessed requests, and the node's partial signature of the current state's hash.
```

Author: I think I'll stick to my version. I think mine is less ambiguous.

> Then nodes have to agree on which batch proposals they want to work on. In short, Alice, Bob, and Charlie have to confirm that they plan to work on proposals from Alice, Bob, and Charlie, and from no one else. As long as there are more than two thirds of honest nodes, they will be able to find an *asynchronous common subset* of the batch proposals. From that point nodes have the same input and will produce the same result independently.

Reviewer: ```suggestion
Then the nodes have to agree on which batch proposals they want to work on. In short, Alice, Bob, and Charlie have to confirm that they plan to work on proposals from Alice, Bob, and Charlie and no one else. As long as there are more than two thirds of honest nodes, they will be able to find an *asynchronous common subset* of the batch proposals. From that point, nodes have the same input and will produce the same result independently.
```
If you plan on using Alice Bob and Charlie as example it  may be a good idea to introduce them or the scenario. In this case I'd probably go with validator A, B, and C  or something like that.

## The Batch

> The next step is to convert the raw list of batch proposals into an actual batch: for that, all requests from all proposals are counted and filtered to produce a single list of requests in a certain order.

Reviewer: ```suggestion
The next step is to convert the raw list of batch proposals into an actual batch: for that, all requests from all proposals are counted and filtered to produce a single list of requests in a particular order.
```

Author: I have rewritten this sentence.

> As all nodes had to sign the same piece of information, the hash of the current state, these partial signatures can be combined into a full, valid signature that is then fed to a pseudo-random function that orders the smart contract requests. Validator nodes cannot neither affect nor predict the final order of requests in the batch.

Reviewer: ```suggestion
As all nodes have to sign the same piece of information, the hash of the current state, these partial signatures can be combined into a full, valid signature that is then fed to a pseudo-random function that orders the smart contract requests. Validator nodes can neither affect nor predict the final order of requests in the batch.
```

Author: Since the signing came from the previous step, I think I'll keep the past tense for the clarity sake.

## State Anchor

> After agreeing on the input, each node executes every smart contract request in order, producing the same new block independently. Each node then crafts a state anchor, a Layer 1 transaction that proves the commitment to this new chain state. The timestamp for this transaction is derived from timestamps of all batch proposals.

Reviewer: ```suggestion
After agreeing on the input, each node executes every smart contract request in order, producing the same new block independently. Each node then crafts a state anchor, a Layer 1 transaction that proves the commitment to this new chain state. The timestamp for this transaction is derived from the timestamps of all batch proposals.
```

> All nodes then sign the state anchor with their partial keys and exchange these signatures. Now all nodes have a valid combined signature and any node can send the anchor transaction to Layer 1. In theory, nodes could publish every single crafted transaction, but in practice ISC committees skip over some state anchors to reduce the load on the Tangle.

Reviewer: ```suggestion
All nodes sign the state anchor with their partial keys and exchange these signatures, so all nodes have a valid combined signature, and any node can send the anchor transaction to Layer 1. In theory, nodes could publish every single transaction. However,  in practice, ISC committees skip over some state anchors to reduce the load on the Tangle.
```
